server/controllers/entitlement/entitlement.go

```diff
@@ -0,0 +1,160 @@
+	subscriptionId string
+	database       *persistence.Database
+}
+
+type APIFilter struct {
+	VendorProductCode   string `json:"vendorProductCode,omitempty"`
+	AzureSubscriptionId string `json:"azureSubscriptionId,omitempty"`
+	AzureTenantId       string `json:"azureTenantId,omitempty"`
+}
+
+type APIResponseContent struct {
+	SourcePartner     string
+	RhAccountId       string
+	Status            string
+	PartnerIdentities map[string]string
+	RhEntitlements    []map[string]string
+}
+
+type APIResponse struct {
+	Content []APIResponseContent
+	Page    map[string]float64
+}
+
+var (
+	once                    sync.Once
+	entitlementCtrlInstance *EntitlementAPIController
+)
+
+func NewEntitlementController(context context.Context, db *persistence.Database) *EntitlementAPIController {
+	once.Do(func() {
+
+		cert := config.GetEnvironment().SW_SUB_API_CERTIFICATE
+		key := config.GetEnvironment().SW_SUB_API_PRIVATEKEY
+		url := config.GetEnvironment().SW_SUB_API_URL
+		code := config.GetEnvironment().SW_SUB_VENDOR_PRODUCT_CODE
+		subs := config.GetEnvironment().SUBSCRIPTION
+		var requester *util.HttpRequester
+
+		if cert == "" || key == "" {
+			log.Warn("Entitlements controller will not be initialized because certificate or key are not provided.")
+		} else {
+			var err error
+			requester, err = util.NewHttpRequesterWithCertificate(cert, key)
+			if err != nil {
+				log.Warnf("Could not initialize entitlements controller. %v\n", err)
+			}
+		}
+
+		entitlementCtrlInstance = &EntitlementAPIController{
+			ctx:            context,
+			httpRequester:  requester,
+			apiUrl:         url,
+			productCode:    code,
+			subscriptionId: subs,
+			database:       db,
+		}
+	})
+	return entitlementCtrlInstance
+}
+
+func (controller *EntitlementAPIController) FetchSubscriptions() {
+	// no need to wait for this one, its not long running and http request uses context
+	go func() {
+		if controller.httpRequester != nil {
+			// in the future we might need to handle pagination... maybe...
+			resp, err := controller.httpRequester.MakeRequestWithJSONBody(
+				controller.ctx,
+				"POST",
+				controller.apiUrl,
+				nil,
+				APIFilter{
+					VendorProductCode:   controller.productCode,
+					AzureSubscriptionId: controller.subscriptionId,
+					//AzureTenantId:       "",
+				},
+			)
+			if err != nil {
+				log.Warnf("Failed to get response from subscription API: %v", err)
+				return
+			}
+
+			response := APIResponse{}
+			if err := json.Unmarshal(resp.Body, &response); err != nil {
+				log.Errorf("Couldn't unmarshal JSON response. %v", err)
+				return
+			}
+
+			if err := storeEntitlements(controller.database, &response); err != nil {
+				log.Errorf("Couldn't persist entitlements in database. %v", err)
+			}
+
+			return
+		}
+		log.Warn("Entitlements can not be fetched, entitlement controller was not initialized.")
+	}()
+}
+
+func storeEntitlements(db *persistence.Database, data *APIResponse) error {
+	if len(data.Content) == 0 {
```

Comment:
No need to distinguish between "didn't fetch" and "fetched but got nothing"?

Reply:
Yes, I'm about to add a field into the model to store an error message to distinguish between error state (didn't fetch) and got nothing.